stko/optimizers/xtb.py

```diff
@@ -496,3 +514,633 @@ def optimize(self, mol):
+            If the CREST run failed to complete.
+
+        """
+
+        if not os.path.exists(output_file):
+            # No simulation has been run.
+            raise CRESTNotStartedError('CREST run did not start.')
+
+        elif any(not os.path.exists(i) for i in output_xyzs):
+            # Best conformer was not output.
+            raise CRESTNotCompletedError('CREST run did not complete.')
+
+        return True
+
+    def _run_crest(self, xyz, out_file):
+        """
+        Run CREST along side GFN-xTB.
+
+        Parameters
+        ----------
+        xyz : :class:`str`
+            The name of the input structure ``.xyz`` file.
+
+        out_file : :class:`str`
+            The name of output file with xTB results.
+
+        Returns
+        -------
+        None : :class:`NoneType`
+
+        """
+
+        # Modify the memory limit.
+        if self._unlimited_memory:
+            memory = 'ulimit -s unlimited ;'
+        else:
+            memory = ''
+
+        # Set optimization level and type.
+        optimization = f'-opt {self._opt_level}'
+        keepdirs = '-keepdir' if self._keepdir is True else ''
+        speed_settings = (
+            f'-{self._speed_setting}'
+            if self._speed_setting is not None else ''
+        )
+
+        cmd = (
+            f'{memory} {self._crest_path} {xyz} '
+            f'-xnam {self._xtb_path} -nozs '
+            f'-ewin {self._ewin} '
+            f'-chrg {self._charge} '
+            f'-gff {speed_settings} '
+            f'{optimization} -T {self._num_cores} {keepdirs}'
+        )
+
+        with open(out_file, 'w') as f:
+            # Note that sp.call will hold the program until completion
+            # of the calculation.
+            sp.call(
+                cmd,
+                stdin=sp.PIPE,
+                stdout=f,
+                stderr=sp.PIPE,
+                # Shell is required to run complex arguments.
+                shell=True
+            )
+
+    def _run_optimization(self, mol):
+        """
+        Run loop of optimizations on `mol` using xTB.
+
+        Parameters
+        ----------
+        mol : :class:`.Molecule`
+            The molecule to be optimized.
+
+        Returns
+        -------
+        mol : :class:`.Molecule`
+            The optimized molecule.
+
+        opt_complete : :class:`bool`
+            Returns ``True`` if the calculation is complete and
+            ``False`` if the calculation is incomplete.
+
+        """
+
+        xyz = f'input_structure_ff.xyz'
+        out_file = f'crest_ff.output'
+        mol.write(xyz)
+        self._run_crest(xyz=xyz, out_file=out_file)
+
+        # Check if the optimization is complete.
+        output_xyzs = [
+            'crest_best.xyz',
+            'crest_conformers.xyz',
+            'crest_rotamers.xyz'
+        ]
+        opt_complete = self._is_complete(out_file, output_xyzs)
+        mol = mol.with_structure_from_file(output_xyzs[0])
```

Comment:
Same comment as above. With will this break if opt_complete is False? I see it's called in a Try, Except clause but is it better to do `if not opt_complete: raise`

Reply:
Agreed, if you look at _is_complete, you will see that if it is incomplete, it raises